### ISSUE TITLE
PDF preview not shown when using \bookOutputName and \bookOutputSuffix.

### ISSUE THREAD

**Issue by rdoursenaud** [wbsoft/frescobaldi#546]
Lilypond 2.18 allows to customize the output filenames with \bookOutputName and \bookOutputSuffix.

Frescobaldi won't display PDF preview when both are used at the same time.

Granted it's not in the LilyPond documentation, but it works.

**Comment by wbsoft** [comment]
Can you post a small example file using both commands and exhibiting this problem? Currently Frescobaldi just looks for both commands, not relating them to a particular score or book. So Frescobaldi is not able to understand which suffix has to be appended to which name. This can be improved of course. We already have infrastructure to locate score, book sections etc and then we could see which suffix applies to which document which also can already have an output name set.

**Comment by wbsoft** [comment]
Just to keep track of my thoughts: LilyPond used to look at the global `output-suffix` variable. But now it also looks at the paper variables `output-filename` and `output-suffix`. If this combination occurs more than once, a number is appended (like `-1`). If the `\bookOutputName` or `\bookOutputSuffix` command occurs at the toplevel, the global paper object is modified and all books are affected. If the command occurs inside a book, only that book is affected.

**Comment by dominicus** [comment]
Frescobaldi 2.0.13 recognizes an explicit string setting for \bookOutputName, but fails to present the preview if a variable is used.
```
%**** save this code with any filename
%**** other than "example.ly"
\version "2.18.2"
filename = "example"
\bookOutputName \filename
{ c''}
```

**Comment by wbsoft** [comment]
Yes that would really require parsing both LilyPond and scheme. We could try but we'll fail with things like

```lilypond
#(define outputsuf "foo")
\bookOutputName #(string-append outputsuf "there")
```

**Comment by sincere-music** [comment]
I also happen to be impeded by this if I try to call \bookOutputName from inside any function. Couldn’t Frescobaldi parse the effective output file names from the LilyPond log and include all of them in the file name drop-down menu for the PDF preview?

**Comment by wbsoft** [comment]
That's a VERY clever idea!

**Comment by pl-gauthier** [comment]
This is a minimal working example showing that, in this case, Frescobaldi is not aware a PDF was generated.
```
\version "2.19.46"

#(ly:book-process
  (ly:make-book
   $defaultpaper
   $defaultheader
   #{ \score {c'} #})
  $defaultpaper
  $defaultlayout
  "ThisIsThePDFName")
```
And, as @sincere-music said, I think getting LilyPond to explicitly **tell** you what file it actually generated seems a more elegant way than to reverse engineer LilyPonds internals and the music files to guess what its output will be.

**Comment by pl-gauthier** [comment]
Would it be a workaround to be able to manually add PDF files to the `Generated Files` list?

**Comment by wbsoft** [comment]
You can use the `output` document variable to tell Frescobaldi which output prefix documents are to be looked for.

But indeed: it would be fantastic if LilyPond could provide the names of generated output files in the log (during processing) in a parseable way.

**Comment by sincere-music** [comment]
Am 03.08.2016 um 14:04 schrieb Wilbert Berendsen: 

You can use the output document variable to tell Frescobaldi which output prefix documents are to be looked for. 

But indeed: it would be fantastic if LilyPond could provide the names of generated output files in the log (during processing) in a parseable way.  
 Oh, I imagined this part would be stereotypical enough for parsing:

 Layout output to `test.ps'...
 Converting to `./test.pdf'...

 If it isn't (sorry, I know nothing about implementing such a thing), we should definitely raise a LilyPond issue to make it parseable. Could you then give any more information on what that would require, please?

**Comment by uliska** [comment]
That was my idea as well. But then Frescobaldi would have to parse all conceivable languages ...

To make it parsable it shouldn't be dependent on the log language.

**Comment by sincere-music** [comment]
Might it perhaps suffice to parse for any occurrence of
```
`*.pdf'...
```
? I don’t think that this pattern can occur anywhere else in the log.

**Comment by pl-gauthier** [comment]
Is there a way to manually select some `pdf` files to be shown in the `Music View` ?

@wbsoft said :
> You can use the output document variable to tell Frescobaldi which output prefix documents are to be looked for.

I am not sure what you mean.
Should the output variable be set *in* the lilypond code ? 
or in the lilypond command itself (i.e. lilypond --output "stuff") ? 
or in Frescobaldi ?

**Comment by pl-gauthier** [comment]
I just realized I could open a `pdf` file *in* the editor hence showing the binary in the editor window which is a little slow to load. It's only when this file is loaded that I can access the `pdf` _(clicking, etc)_ in the `Music View`.

**Comment by pl-gauthier** [comment]
Given that:

- By introspecting some to-be-compiled LilyPond files, one cannot reliably infer the pdf filenames without having to code LilyPond itself ;
- The LilyPond output cannot be parsed reliably (as Urs mentioned).
- I think it is not Frescobaldi's responsibility to guess what LilyPond will do but rather LilyPond to express clearly what it is it made.

I think a solution could be for LilyPond to produce a manifest file at compilation.

The manifest file could be set as an option at compilation just like the temporary folder is by Frescobaldi so that Frescobaldi could reliably find said manifest. Frescobaldi could then read that manifest at the end of compilation.

Obviously this compiler manifest should be plain text and in a easily digestible format (e.g. one file per line).

    ./music.pdf
    ./music.svg
    ./music.midi
